src/audio/mixin_mixout.c

```diff
@@ -248,235 +263,367 @@ struct mixout_source_info *find_mixout_source_info(struct mixed_data_info __spar
+	 * division operations.
+	 */
+	while (frames_to_mix) {
+		src = audio_stream_wrap(source, src);
+		dest = audio_stream_wrap(sink, dest);
+
+		if (gain == IPC4_MIXIN_UNITY_GAIN)
+			while (frames_to_mix && src < (int32_t *)source->end_addr &&
+			       dest < (int32_t *)sink->end_addr) {
 				*dest = sat_int24(sign_extend_s24(*dest) + sign_extend_s24(*src));
-				src++;
-				dest++;
+				src += source_channel_count;
+				dest += sink_channel_count;
+				frames_to_mix--;
 			}
-		} else {
-			for (i = 0; i < n; i++) {
+		else
+			while (frames_to_mix && src < (int32_t *)source->end_addr &&
+			       dest < (int32_t *)sink->end_addr) {
 				*dest = sat_int24(sign_extend_s24(*dest) +
-				  (int32_t)q_mults_32x32(sign_extend_s24(*src),
-							 gain, IPC4_MIXIN_GAIN_SHIFT));
-				src++;
-				dest++;
+						  (int32_t)q_mults_32x32(sign_extend_s24(*src),
+						  gain, IPC4_MIXIN_GAIN_SHIFT));
+				src += source_channel_count;
+				dest += sink_channel_count;
+				frames_to_mix--;
 			}
-		}
+	}
 
-		dest = audio_stream_wrap(sink, dest);
+	while (frames_to_copy) {
 		src = audio_stream_wrap(source, src);
-	}
+		dest = audio_stream_wrap(sink, dest);
 
-	if (gain == IPC4_MIXIN_UNITY_GAIN) {
-		audio_stream_bytes_copy(sink, dest, source, src, bytes_to_copy);
-	} else {
-		while (bytes_to_copy != 0) {
-			n = MIN(audio_stream_bytes_without_wrap(sink, dest), bytes_to_copy);
-			n = MIN(audio_stream_bytes_without_wrap(source, src), n);
-			bytes_to_copy -= n;
-			n >>= 2;	/* divide by 4 */
-			for (i = 0; i < n; i++) {
+		if (gain == IPC4_MIXIN_UNITY_GAIN)
+			while (frames_to_copy && src < (int32_t *)source->end_addr &&
+			       dest < (int32_t *)sink->end_addr) {
+				*dest = *src;
+				src += source_channel_count;
+				dest += sink_channel_count;
+				frames_to_copy--;
+			}
+		else
+			while (frames_to_copy && src < (int32_t *)source->end_addr &&
+			       dest < (int32_t *)sink->end_addr) {
 				*dest = (int32_t)q_mults_32x32(sign_extend_s24(*src),
 							       gain, IPC4_MIXIN_GAIN_SHIFT);
-				src++;
-				dest++;
+				src += source_channel_count;
+				dest += sink_channel_count;
+				frames_to_copy--;
 			}
-
-			dest = audio_stream_wrap(sink, dest);
-			src = audio_stream_wrap(source, src);
-		}
 	}
 }
+
 #endif	/* CONFIG_FORMAT_S24LE */
 
 #if CONFIG_FORMAT_S32LE
-static void mix_s32(struct audio_stream __sparse_cache *sink, uint32_t start, uint32_t mixed_bytes,
-		    const struct audio_stream __sparse_cache *source, uint32_t size, uint16_t gain)
+/* Instead of using sink->channels and source->channels, sink_channel_count and
+ * source_channel_count are supplied as parameters. This is done to reuse the function to also
+ * mix entire stream. In this case the function is called with fake stream parameters: multichannel
+ * stream is represented as single channel and so the entire stream contents is mixed.
+ */
+static void mix_channel_s32(struct audio_stream __sparse_cache *sink, uint8_t sink_channel_index,
```

Comment:
Good idea to reuse, but I would rename the variables that are used only for remapping with a ```remap_``` prefix and make it obvious in the comment that remapping is enabled by setting them to non zero values.

Reply:
@lgirdwood, I do not understand your suggestion. In channel remapping mode channels are mixed using mix_channel_sXX(); in normal mode, multichannel stream is treated as a single channel and so an entire stream is mixed using the same mix_channel_sXX() function dedicated to channel mixing. That function is unaware if it works on real channel of a multichannel stream or on a fake channel (a multichannel stream represented as single channel).
What variables do you suggest to mark with remap_ prefix?

Reply:
The part that confuses me is the work "fake". I can also see 2 modes, NORMAL and REMAP (where you hard code 0, 1 in one of the modes). I would like to see it clear in the comment if you use mode X then you must set x,y = 0,1. This way everyone will understand how to use it.

Reply:
Added comment in mix_and_remap() how an entire stream is mixed reusing mix_channel() function.